source/Halibut/Transport/Protocol/MessageExchangeProtocol.cs

```diff
@@ -41,24 +41,29 @@ public ResponseMessage ExchangeAsClient(RequestMessage request)
 
         public async Task<ResponseMessage> ExchangeAsClientAsync(RequestMessage request, CancellationToken cancellationToken)
         {
-            // TODO - ASYNC ME UP!
-            await Task.CompletedTask;
+            await PrepareExchangeAsClientAsync(cancellationToken);
 
-#pragma warning disable CS0612
-            return ExchangeAsClient(request);
-#pragma warning restore CS0612
+            await stream.SendAsync(request, cancellationToken);
+            return await stream.ReceiveAsync<ResponseMessage>(cancellationToken);
         }
 
         public void StopAcceptingClientRequests()
         {
             acceptClientRequests = false;
         }
 
+        [Obsolete]
         public void EndCommunicationWithServer()
         {
             stream.SendEnd();
         }
 
+        public async Task EndCommunicationWithServer(CancellationToken cancellationToken)
```

Comment:
Should this end in Async for consistency.

Reply:
Yes it should 👍